getResources can include non-child resources

### Describe the bug

In PR #571 , a validation was added to exclude non-child resources, however the validation misses some edge cases.

### Expected behavior

`getResources()` from a container `/foo/` should not include `/foo/#bar` as a child (and anything else that is not a direct child resource).

### Actual behavior

It includes `/foo/#bar` as a child

### How to reproduce?

I've tested it with these lines in container.ttl and the unit tests:

```text/turtle
# These containment triples should not be included in a getResources response
<>
    ldp:contains <https://example.com/other> , <newContainer/child> , <> , <./>, <child/..>, <#hashtag>, <../> .
```

The test case failed because the hashtag was not filtered:

```
[ERROR]   SolidClientTest.testLowLevelSolidContainer » Completion org.opentest4j.AssertionFailedError: expected: <[http://localhost:36023/container/newContainer/, http://localhost:36023/container/test.txt, http:
//localhost:36023/container/test2.txt]> but was: <[http://localhost:36023/container/newContainer/, http://localhost:36023/container/test.txt, http://localhost:36023/container/test2.txt, http://localhost:36023/co
ntainer/#hashtag]>
```

### Output of `java -version`

_No response_

### Maven version

_No response_

### Additional information

### Cause

The problem likely is, that in the verifyContainmentIri method, the `normalizedPath.isEmpty()` is called as `"#hashtag".isEmpty()`. In other words, it thinks that the "#hashtag" is a child resource, rather than just a hashtag.

### Suggested fix

I'm not aware how to do this in Java, but I'd leave the path parsing to a library that already has all the edge cases built in (hashtag, query params, and not sure if there's more). Similar to how in solid-client-js they use the built-in `child.pathname` rather than manually parsing the path segments.

--- a/solid/src/main/java/com/inrupt/client/solid/SolidContainer.java
+++ b/solid/src/main/java/com/inrupt/client/solid/SolidContainer.java
@@ -82,9 +82,9 @@ public class SolidContainer extends SolidRDFSource {
      * @return the contained resources
      */
     public Set<SolidResource> getResources() {
-        final String container = normalize(getIdentifier());
         // As defined by the Solid Protocol, containers always end with a slash.
-        if (container.endsWith("/")) {
+        if (getIdentifier().getPath().endsWith("/")) {
+            final String container = normalize(getIdentifier());
             final Node node = new Node(rdf.createIRI(getIdentifier().toString()), getGraph());
             try (final Stream<Node.TypedNode> stream = node.getResources()) {
                 return stream.filter(child -> verifyContainmentIri(container, child)).map(child -> {
@@ -100,14 +100,14 @@ public class SolidContainer extends SolidRDFSource {
 
     @Override
     public ValidationResult validate() {
-        // Get the normalized container URI
-        final String container = normalize(getIdentifier());
         final List<String> messages = new ArrayList<>();
         // Verify that the container URI path ends with a slash
-        if (!container.endsWith("/")) {
+        if (!getIdentifier().getPath().endsWith("/")) {
             messages.add("Container URI does not end in a slash");
         }
 
+        // Get the normalized container URI
+        final String container = normalize(getIdentifier());
         // Verify that all ldp:contains triples align with Solid expectations
         getGraph().stream(null, rdf.createIRI(LDP.contains.toString()), null)
             .collect(Collectors.partitioningBy(verifyContainmentTriple(container)))
@@ -119,10 +119,6 @@ public class SolidContainer extends SolidRDFSource {
             return new ValidationResult(true);
         }
         return new ValidationResult(false, messages);
-    }
-
-    static String normalize(final IRI iri) {
-        return normalize(URI.create(iri.getIRIString()));
     }
 
     static String normalize(final URI uri) {
@@ -145,22 +141,43 @@ public class SolidContainer extends SolidRDFSource {
     }
 
     static boolean verifyContainmentIri(final String container, final IRI object) {
-        if (!object.getIRIString().startsWith(container)) {
-            // Out-of-domain containment triple object
+
+        // 1 URI Structure Tests
+        final URI normalized = URI.create(object.getIRIString()).normalize();
+
+        // 1.A Query strings are not allowed in object URI
+        if (normalized.getQuery() != null) {
             return false;
-        } else {
-            final String relativePath = object.getIRIString().substring(container.length());
-            final String normalizedPath = relativePath.endsWith("/") ?
-                relativePath.substring(0, relativePath.length() - 1) : relativePath;
-            if (normalizedPath.isEmpty()) {
-                // Containment triple subject and object cannot be the same
-                return false;
-            }
-            if (normalizedPath.contains("/")) {
-                // Containment cannot skip intermediate nodes
-                return false;
-            }
         }
+
+        // 1.B URI fragments are not allowed in object URI
+        if (normalized.getFragment() != null) {
+            return false;
+        }
+
+        // 2 Relative path tests
+        final URI base = URI.create(container).normalize();
+        final URI relative = base.relativize(normalized);
+
+        // 2.A Base URI cannot equal the object URI
+        if (base.equals(normalized)) {
+            return false;
+        }
+
+        // 2.B Object URI must be relative to (contained in) the base URI
+        if (relative.isAbsolute()) {
+            return false;
+        }
+
+        final String relativePath = relative.getPath();
+        final String normalizedPath = relativePath.endsWith("/") ?
+            relativePath.substring(0, relativePath.length() - 1) : relativePath;
+
+        // 2.C Containment cannot skip intermediate nodes
+        if (normalizedPath.contains("/")) {
+            return false;
+        }
+
         return true;
     }
 

--- a/solid/src/main/java/com/inrupt/client/solid/SolidContainer.java
+++ b/solid/src/main/java/com/inrupt/client/solid/SolidContainer.java
@@ -83,9 +83,10 @@ public class SolidContainer extends SolidRDFSource {
      */
     public Set<SolidResource> getResources() {
         // As defined by the Solid Protocol, containers always end with a slash.
-        if (getIdentifier().getPath().endsWith("/")) {
-            final String container = normalize(getIdentifier());
-            final Node node = new Node(rdf.createIRI(getIdentifier().toString()), getGraph());
+        final URI base = getIdentifier().normalize();
+        if (isContainer(base)) {
+            final String container = normalize(base);
+            final Node node = new Node(rdf.createIRI(base.toString()), getGraph());
             try (final Stream<Node.TypedNode> stream = node.getResources()) {
                 return stream.filter(child -> verifyContainmentIri(container, child)).map(child -> {
                     final Metadata.Builder builder = Metadata.newBuilder();
@@ -102,12 +103,13 @@ public class SolidContainer extends SolidRDFSource {
     public ValidationResult validate() {
         final List<String> messages = new ArrayList<>();
         // Verify that the container URI path ends with a slash
-        if (!getIdentifier().getPath().endsWith("/")) {
+        final URI base = getIdentifier().normalize();
+        if (!isContainer(base)) {
             messages.add("Container URI does not end in a slash");
         }
 
         // Get the normalized container URI
-        final String container = normalize(getIdentifier());
+        final String container = normalize(base);
         // Verify that all ldp:contains triples align with Solid expectations
         getGraph().stream(null, rdf.createIRI(LDP.contains.toString()), null)
             .collect(Collectors.partitioningBy(verifyContainmentTriple(container)))
@@ -119,6 +121,10 @@ public class SolidContainer extends SolidRDFSource {
             return new ValidationResult(true);
         }
         return new ValidationResult(false, messages);
+    }
+
+    static boolean isContainer(final URI uri) {
+        return uri.normalize().getPath().endsWith("/");
     }
 
     static String normalize(final URI uri) {
@@ -142,29 +148,30 @@ public class SolidContainer extends SolidRDFSource {
 
     static boolean verifyContainmentIri(final String container, final IRI object) {
 
-        // 1 URI Structure Tests
+        // URI Structure Tests
+        final URI base = URI.create(container).normalize();
         final URI normalized = URI.create(object.getIRIString()).normalize();
 
-        // 1.A Query strings are not allowed in object URI
-        if (normalized.getQuery() != null) {
+        // Query strings are not allowed in subject or object URI
+        if (base.getQuery() != null || normalized.getQuery() != null) {
             return false;
         }
 
-        // 1.B URI fragments are not allowed in object URI
-        if (normalized.getFragment() != null) {
+        // URI fragments are not allowed in subject or object URI
+        if (base.getFragment() != null || normalized.getFragment() != null) {
             return false;
         }
 
-        // 2 Relative path tests
-        final URI base = URI.create(container).normalize();
+        // Base URI cannot equal the object URI
+        if (base.getScheme().equals(normalized.getScheme()) &&
+                base.getSchemeSpecificPart().equals(normalized.getSchemeSpecificPart())) {
+            return false;
+        }
+
+        // Relative path tests
         final URI relative = base.relativize(normalized);
 
-        // 2.A Base URI cannot equal the object URI
-        if (base.equals(normalized)) {
-            return false;
-        }
-
-        // 2.B Object URI must be relative to (contained in) the base URI
+        // Object URI must be relative to (contained in) the base URI
         if (relative.isAbsolute()) {
             return false;
         }
@@ -173,7 +180,7 @@ public class SolidContainer extends SolidRDFSource {
         final String normalizedPath = relativePath.endsWith("/") ?
             relativePath.substring(0, relativePath.length() - 1) : relativePath;
 
-        // 2.C Containment cannot skip intermediate nodes
+        // Containment cannot skip intermediate nodes
         if (normalizedPath.contains("/")) {
             return false;
         }

--- a/solid/src/test/java/com/inrupt/client/solid/SolidClientTest.java
+++ b/solid/src/test/java/com/inrupt/client/solid/SolidClientTest.java
@@ -231,6 +231,8 @@ class SolidClientTest {
         expected.add(URIBuilder.newBuilder(uri).path("newContainer/").build());
         expected.add(URIBuilder.newBuilder(uri).path("test.txt").build());
         expected.add(URIBuilder.newBuilder(uri).path("test2.txt").build());
+        expected.add(URIBuilder.newBuilder(uri).path("test3").build());
+        expected.add(URIBuilder.newBuilder(uri).path("test4").build());
 
         client.send(Request.newBuilder(uri).build(), SolidResourceHandlers.ofSolidContainer())
             .thenAccept(response -> {

--- a/solid/src/test/java/com/inrupt/client/solid/SolidRDFSourceTest.java
+++ b/solid/src/test/java/com/inrupt/client/solid/SolidRDFSourceTest.java
@@ -195,6 +195,25 @@ class SolidRDFSourceTest {
     }
 
     @Test
+    void testNonContainer() {
+        final URI resource = URI.create(config.get("solid_resource_uri") + "/recipe");
+        final Request request = Request.newBuilder()
+            .uri(resource)
+            .header("Accept", "text/turtle")
+            .GET()
+            .build();
+
+        final Response<SolidContainer> response = client.send(request, SolidResourceHandlers.ofSolidContainer())
+            .toCompletableFuture().join();
+
+        try (final SolidContainer container = response.body()) {
+            assertEquals(resource, container.getIdentifier());
+            assertTrue(container.getResources().isEmpty());
+            assertFalse(container.validate().isValid());
+        }
+    }
+
+    @Test
     void testInvalidRdf() {
         final URI resource = URI.create(config.get("solid_resource_uri") + "/nonRDF");
         final Request request = Request.newBuilder()

--- a/solid/src/test/resources/__files/container.ttl
+++ b/solid/src/test/resources/__files/container.ttl
@@ -7,6 +7,14 @@
     a ldp:BasicContainer ;
     dct:modified "2022-11-25T10:36:36Z"^^xsd:dateTime;
     ldp:contains <newContainer/>, <test.txt>, <test2.txt> .
+<intermediate/..>
+    a ldp:BasicContainer ;
+    dct:modified "2022-11-25T10:38:12Z"^^xsd:dateTime ;
+    ldp:contains <test3> .
+<intermediate/../>
+    a ldp:BasicContainer ;
+    dct:modified "2022-11-25T10:38:47Z"^^xsd:dateTime ;
+    ldp:contains <test4> .
 <newContainer/>
     a ldp:BasicContainer ;
     dct:modified "2022-11-25T10:36:36Z"^^xsd:dateTime .
@@ -16,10 +24,16 @@
 <test2.txt>
     a pl:Resource, ldp:NonRDFSource;
     dct:modified "2022-11-25T10:37:06Z"^^xsd:dateTime .
+<test3>
+    a ldp:RDFSource ;
+    dct:modified "2022-11-25T10:37:31Z"^^xsd:dateTime .
+<test4>
+    a ldp:RDFSource ;
+    dct:modified "2022-11-25T10:39:22Z"^^xsd:dateTime .
 
 # These containment triples should not be included in a getResources response
 <>
-    ldp:contains <https://example.com/other> , <newContainer/child> , <> , <./> ,
+    ldp:contains <https://example.com/other> , <newContainer/child> , <newContainer%2Fchild2> ,  <> , <./> ,
     <?foo> , <#bar> , <?foo#bar> , <./?foo> , <./#bar> , <./?foo#bar> ,
     <../?foo> , <../#bar> , <../?foo#bar> , <child?foo> , <child#bar> , <child?foo#bar> .
 <https://example.test/container/>

--- a/solid/src/test/resources/__files/container.ttl
+++ b/solid/src/test/resources/__files/container.ttl
@@ -19,7 +19,9 @@
 
 # These containment triples should not be included in a getResources response
 <>
-    ldp:contains <https://example.com/other> , <newContainer/child> , <> , <./> .
+    ldp:contains <https://example.com/other> , <newContainer/child> , <> , <./> ,
+    <?foo> , <#bar> , <?foo#bar> , <./?foo> , <./#bar> , <./?foo#bar> ,
+    <../?foo> , <../#bar> , <../?foo#bar> , <child?foo> , <child#bar> , <child?foo#bar> .
 <https://example.test/container/>
     a ldp:BasicContainer ;
     ldp:contains <https://example.test/container/external> .

